hbase-server/src/test/java/org/apache/hadoop/hbase/security/provider/TestCustomSaslAuthenticationProvider.java

```diff
@@ -527,12 +568,22 @@ public void testNegativeAuthentication() throws Exception {
     user1.addToken(createPasswordToken("user1", "definitely not the password", clusterId));
     user1.doAs(new PrivilegedExceptionAction<Void>() {
       @Override public Void run() throws Exception {
+        // Depending on the registry in use, the following code can throw exceptions at different
+        // places. Master registry fails at the createConnection() step because the RPC to the
+        // master fails with sasl auth. With ZK registry, connection creation succeeds (since there
+        // is no RPC to HBase services involved) but the subsequent get() fails. The root cause
+        // should still be a SaslException in both the cases.
         try (Connection conn = ConnectionFactory.createConnection(clientConf);
             Table t = conn.getTable(tableName)) {
           t.get(new Get(Bytes.toBytes("r1")));
           fail("Should not successfully authenticate with HBase");
-          return null;
+        } catch (MasterRegistryFetchException mfe) {
+          Throwable cause = mfe.getCause();
+          assertTrue(cause.getMessage(), cause.getMessage().contains("SaslException"));
+        } catch (RetriesExhaustedException re) {
+          assertTrue(re.getMessage(), re.getMessage().contains("SaslException"));
         }
+        return null;
```

Comment:
There's a `throws Exception` on `run()` so we need  a `fail()` "close" to the `return null`. Right now, an exception other than MasterRegistryFetchException and RetriesExhaustedException would not cause a unit test failure.

You can just move the `fail` from inside the try, next to this `return null` and change the message to be something like `Expected an authentication failure`.

Reply:
> Right now, an exception other than MasterRegistryFetchException and RetriesExhaustedException would not cause a unit test failure.

Not sure I follow you, the exception should be propagated back to the test (since it is not caught) and it should fail, no? I tried the following it works (meaning the IAE is thrown by the unit test runner). Did I miss something?

 ```
user1.doAs(new PrivilegedExceptionAction<Void>() {
      @Override public Void run() throws Exception {
        // Depending on the registry in use, the following code can throw exceptions at different
        // places. Master registry fails at the createConnection() step because the RPC to the
        // master fails with sasl auth. With ZK registry, connection creation succeeds (since there
        // is no RPC to HBase services involved) but the subsequent get() fails. The root cause
        // should still be a SaslException in both the cases.
        try (Connection conn = ConnectionFactory.createConnection(clientConf);
            Table t = conn.getTable(tableName)) {
          if (true) {
            throw new IllegalArgumentException("test");
          }
          t.get(new Get(Bytes.toBytes("r1")));

java.lang.IllegalArgumentException: test

	at org.apache.hadoop.hbase.security.provider.TestCustomSaslAuthenticationProvider$3.run(TestCustomSaslAuthenticationProvider.java:579)
	at org.apache.hadoop.hbase.security.provider.TestCustomSaslAuthenticationProvider$3.run(TestCustomSaslAuthenticationProvider.java:569)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
```

Reply:
You're right that the thrown, uncaught exception will ultimately cause the test to fail, but you're removing the nice exception stating the reason the test failed :)

Reply:
Okay, added back a nice exception. The issue with moving fail() towards the end is that we need sprinkle the "return null" statements in the catch blocks. This is more readable IMO.

Reply:
Yup, what you did is fine. Thanks!